python/pyspark/ml/torch/distributor.py

```diff
@@ -0,0 +1,491 @@
+    args: *args
+        The input arguments to the train.py file.
+
+    Returns
+    -------
+    str
+        The output torchrun command
+    """
+    local_mode = input_params["local_mode"]
+    num_processes = input_params["num_processes"]
+
+    if local_mode:
+        standalone = ["--standalone", "--nnodes=1"]
+        processes_per_node = num_processes
+    else:
+        master_addr, master_port = os.environ["MASTER_ADDR"], os.environ["MASTER_PORT"]
+        node_rank = os.environ["RANK"]
+        standalone = [
+            f"--nnodes={num_processes}",
+            f"--node_rank={node_rank}",
+            f"--rdzv_endpoint={master_addr}:{master_port}",
+            "--rdzv_id=0",
+        ]  # TODO: setup random ID that is gleaned from env variables
+        processes_per_node = 1
+
+    args_string = list(map(str, args))  # converting all args to strings
+
+    return (
+        ["torchrun"]
+        + standalone
+        + [f"--nproc_per_node={processes_per_node}"]
+        + [train_path, *args_string]
+    )
+
+
+def execute_command(cmd: list[str], _prctl: bool = True, redirect_to_stdout: bool = True) -> None:
+    """Runs a command in a new process, redirects stdout/stderr to stdout, and handles termination
+    of the command process.
+
+    Parameters
+    ----------
+    cmd : list[str]
+        The users's input command split by the space characeter
+    _prctl : bool, optional
+        (test only) Use prctl to signal the command process upon parent death, by default True
+    redirect_to_stdout : bool, optional
+        whether to redirect all stdout/stderr logs to current stdout, by default True
+    TODO(SPARK-41593): Will update with additional parameters for logging to driver
+
+    Raises
+    ------
+    RuntimeError
+        If the child executed command fails
+    """
+    _TAIL_LINES_TO_KEEP = 100
+
+    def sigterm_on_parent_death() -> None:
+        """
+        Uses prctl to automatically send SIGTERM to the command process when its parent is dead.
+        This handles the case when the parent is a PySpark worker process.
+        If a user cancels the PySpark job, the worker process gets killed, regardless of
+        PySpark daemon and worker reuse settings.
+        NOTE: This doesn't work on windows or POSIX.
+        Future reference:
+        - MacOS/POSIX: https://stackoverflow.com/a/2035683
+        - Windows: https://stackoverflow.com/a/23587108
+        """
+        if _prctl:
+            try:
+                libc = ctypes.CDLL("libc.so.6")
+                # Set the parent process death signal of the command process to SIGTERM.
+                libc.prctl(1, signal.SIGTERM)
+            except OSError:
+                pass
+
+    task = subprocess.Popen(
+        cmd,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.STDOUT,
+        stdin=subprocess.PIPE,
+        env=os.environ,  # preserve the environ to support nested MPI jobs
+        preexec_fn=sigterm_on_parent_death,
```

Comment:
I think checking parent pid becomes 1 will be a better solution for parent process died event.

Reply:
I agree, but I am not sure how to do that. Since we are calling `torchrun` with `subprocess.Popen([torchrun,...]...)`, where do we have the thread checking for parent id?

Reply:
It's a good question. We can create a wrapping script to do that, the wrapping script creates `torchrun` subprocess, and spark code creates wrapping script subprocess.
We can do it later when we need to support all Unix system

Reply:
Okay I will create a ticket for this.

Reply:
The temporary commit for how this will be done is https://github.com/rithwik-db/spark_pytorch_distributor/commit/39ad2b76301e710283b41008ba63b1a9c5025ebe